src/block_cache/block_tier.c

```diff
@@ -0,0 +1,137 @@
+/*-
+ * Copyright (c) 2014-present MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+
+/*
+ * __wt_blkcache_tiered_open --
+ *     Open a tiered object.
+ */
+int
+__wt_blkcache_tiered_open(
+  WT_SESSION_IMPL *session, const char *uri, uint32_t objectid, WT_BLOCK **blockp)
+{
+    WT_BLOCK *block;
+    WT_BUCKET_STORAGE *bstorage;
+    WT_CONFIG_ITEM pfx;
+    WT_DECL_ITEM(tmp);
+    WT_DECL_RET;
+    WT_TIERED *tiered;
+    const char *cfg[2], *object_name, *object_uri, *object_val;
+    bool exist, local_only, readonly;
+
+    *blockp = NULL;
+
+    tiered = (WT_TIERED *)session->dhandle;
+    object_uri = object_val = NULL;
+
+    WT_ASSERT(session, objectid <= tiered->current_id);
+    WT_ASSERT(session, uri == NULL || WT_PREFIX_MATCH(uri, "tiered:"));
+    WT_ASSERT(session, (uri == NULL && objectid != 0) || (uri != NULL && objectid == 0));
+
+    /*
+     * First look for the local file. This will be the fastest access and we retain recent objects
+     * in the local database for awhile. If we're passed a name to open, then by definition it's a
+     * local file.
+     *
+     * FIXME-WT-7590 we will need some kind of locking while we're looking at the tiered structure.
+     * This can be called at any time, because we are opening the objects lazily.
+     */
+    if (uri != NULL)
+        objectid = tiered->current_id;
+    if (objectid == tiered->current_id) {
+        local_only = true;
+        object_uri = tiered->tiers[WT_TIERED_INDEX_LOCAL].name;
+        object_name = object_uri;
+        WT_PREFIX_SKIP_REQUIRED(session, object_name, "file:");
+        readonly = false;
+    } else {
+        local_only = false;
+        WT_ERR(
+          __wt_tiered_name(session, &tiered->iface, objectid, WT_TIERED_NAME_OBJECT, &object_uri));
+        object_name = object_uri;
+        WT_PREFIX_SKIP_REQUIRED(session, object_name, "object:");
+        readonly = true;
+    }
+
+    /* Get the object's configuration. */
+    WT_ERR(__wt_metadata_search(session, object_uri, (char **)&object_val));
+    cfg[0] = object_val;
+    cfg[1] = NULL;
+
+    /* Check if the object exists. */
+    exist = true;
+    if (!local_only)
+        WT_ERR(__wt_fs_exist(session, object_name, &exist));
+    if (exist)
+        WT_ERR(
+          __wt_block_open(session, object_name, objectid, cfg, false, readonly, false, 0, &block));
```

Comment:
It seems possible that between the existence check and opening the local file on an older object that the object could have been locally dropped (i.e. add a delay after line 70). I think if `wt_block_open` returns an error in the `!local_only` case (i.e. the local file disappeared) then instead of an error the code should try opening in the bucket.

Reply:
This is what @ddanderson was talking about on Slack; there has to be some locking or other verification during open in order for the open and the remove not to race. As far as I know, I've not made this better or worse, it's as broken as it always ways.

Reply:
Well, the race didn't really exist before because we were only calling `wt_open` and not checking existence beforehand, so there was no window. In the old code, if we couldn't open the local (but older) object for any reason we tried opening the bucket object.

Reply:
The previous race was an open followed by a remove which will fail on Windows.